Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
+
+#### API review
+
+Intrinsics will extend the API of CoreCLR.  They will need to follow standard API review practices.
+
+### Choice of Arm64 naming conventions
+
+#### Namespaces
+
+`ARM32` and `ARM64` will follow similar namespace conventions.
+
++ `System.Runtime.Intrinsics` is used for type definitions useful across multiple platforms
++ `System.Runtime.Intrinsics.Arm` is used for enumerations shared across `ARM32` and `ARM64` platforms
++ `System.Runtime.Intrinsics.Arm.Base` is recommended for functionality included in all `ARM32` and `ARM64` platforms
+  + Do we need this or could we assume `ARMv8` is minimum supported platform?
+  + Should we assume `ARMv7` or `ARMv5` is minimum supported platform?
++ `System.Runtime.Intrinsics.Arm.ARMv8` is used for `ARMv8` added functionality included in both `ARM32` and `ARM64` platforms 
++ `System.Runtime.Intrinsics.Arm.ARMv8.Arm64` is used for `ARMv8` functionality only included in `ARM64` platform
++ `System.Runtime.Intrinsics.Arm.ARMv8_1....` is used for `ARMv8.1` functionality.  
+  + Use similar sub namespace names for only included in `ARM64` platform ...
++ `System.Runtime.Intrinsics.Arm.XXX` is used for optional features `CRC32`, `AES`, `SHA1` ... 
+  + All extensions which were not required when they were initially introduced
+  + For example, `CRC32` became required with `ARMv8.1`, but since it was introduced as optional in `ARMv8` it would go in its own namespace
+  
+If an extension is only supported on `Arm32`, there is no reason to add a sub namespace.
+
++ `System.Runtime.Intrinsics.Arm.ARMv7` is used for `ARMv7` added functionality not included in `Base`
+
+If an extension is only supported on `Arm64`, there is no reason to add a sub namespace.  Hypothetically:
+
++ `System.Runtime.Intrinsics.Arm.ARMv10` is used for `ARMv10` added functionality (Hypothetically `Arm64` only)
+
+Namespace is chosen based on what could be implemented, not on what is or will be implemented.
+
+#### Names
+
+Intrinsics will be named to describe functionality.  Names will not correspond to specific named assembly instructions.
+
+`IsSupported` will be used in any namespace which may not be implemented on all CoreCLR platforms
+
+### Intrinsic Interface Documentation
+
++ Documentation will be minimal preferring the underlying ARM documentation
++ Namespaces will briefly document corresponding specification
++ Intrinsic methods will briefly document corresponding assembly instruction(s) for each platform
+
+### Test coverage
+
+As intrinsic support is added test coverage must be extended to provide basic testing
+
+### Phased Implementation
+
+#### Implementation Priorities
+
+As guidelines:
+
++ Baseline functionality will be prioritized over architectural extensions
++ Architectural extensions will typically be prioritized in age order.  Earlier extensions will be added first
++ `ARM64` only and `ARM*` features will have equal priority
++ Priorities will be driven by optimization efforts and requests
```

Comment:
IMHO Arm intrinsics do not seem to be as large as Intel ones. Probably there is no need to go that way but rather implement them in large 2 - 3 chunks at maximum. Otherwise we risk getting to a point where Intel intrinsics were with multiple API Reviews needed to get simple functionality i.e. POPCNT, LZCNT, shifts or shuffles. The logic behind implementing whole sets lies in their design which was made in a way that whole set enables optimal implementation expressiveness. Removing some of the instructions will always create headaches for some implementers.